app/src/main/kotlin/com/wire/android/ui/home/conversations/ConversationViewModel.kt

```diff
@@ -469,18 +473,44 @@ class ConversationViewModel @Inject constructor(
         }
     }
 
-    private suspend fun fromMessageModelToMessageContent(message: Message): MessageContent? =
+    private fun MemberDetails.toSystemMessageName(lowerCaseSelfLabel: Boolean = false) = when (this) {
+        is MemberDetails.Other -> this.name?.let { UIText.DynamicString(it) } ?: UIText.StringResource(R.string.member_name_deleted_label)
+        is MemberDetails.Self -> UIText.StringResource(
+            if (lowerCaseSelfLabel) R.string.member_name_you_label_lowercase
+            else R.string.member_name_you_label_titlecase
+        )
+    }
+
+    private suspend fun fromMessageModelToMessageContent(message: Message, members: List<MemberDetails>): MessageContent? =
         when (message.visibility) {
             Message.Visibility.VISIBLE -> when (val content = message.content) {
                 is Asset -> mapToMessageUI(content.value, message.conversationId, message.id)
                 is Text -> TextMessage(messageBody = MessageBody(UIText.DynamicString(content.value)))
+                is Server -> mapToServerMessageContent(content, message.senderUserId, members)
                 else -> TextMessage(messageBody = MessageBody((content as? Text)?.let { UIText.DynamicString(it.value) }
                     ?: UIText.StringResource(R.string.content_is_not_available)))
             }
             Message.Visibility.DELETED -> DeletedMessage
             Message.Visibility.HIDDEN -> DeletedMessage
         }
 
+    private fun mapToServerMessageContent(content: Server, senderUserId: UserId, members: List<MemberDetails>): MessageContent {
+        return when (content) {
+            is MemberChange -> {
+                val sender = members.findUser(senderUserId)
+                val isAuthorSelfAction = content.members.size == 1 && senderUserId == content.members.first().id
+                val authorName = sender.toSystemMessageName()
+                val memberNameList = content.members.map { members.findUser(it.id).toSystemMessageName(true) }
+                when (content) {
+                    is MemberChange.Join -> ServerMessage.MemberAdded(authorName, memberNameList)
+                    is MemberChange.Leave ->
+                        if(isAuthorSelfAction) ServerMessage.MemberLeft(authorName)
+                        else ServerMessage.MemberRemoved(authorName, memberNameList)
+                }
+            }
+        }
+    }
+
```

Comment:
The same also for this one

Reply:
Also, I think it's better to remove the `map` from the name and make the name simpler.

Reply:
Sure, done (more info in the comment above)